{% extends "layout_help.html" %}
{% block content %}
    <h2>Overview</h2>

    <div class="line"></div>

    <section id="software-arch">
        <h3>Software Architecture</h3>
        <img src="{{url_for('static', filename='help_files/software_arch.png')}}" align="middle" />
        <p style="font-weight: bold;"> Figure 1: POPgen software schematic and component integration </p>
        <p> Figure 1. shows the finalized software structure and all the integrated components used to create the functional POPgen web application, which can search through SNP information within human chromosome 22. This SNP information was downloaded from the IGSR web application and constructed into a functional database using SQLite3, which was then linked to flask using the package SQLAlchemy. SQLAlchemy facilitates communication between the flask python program and the constructed relational SQLite3 database in python. The main application runs on Flask using Jinja2 as a template engine to create HTML and other markup formats which can then be returned to the user in a http response (routes.py is the executable file). Website routes were defined using the HTML language and static files such as CSS (maintain consistency in website design). Together the application renders the final site that also features a user download for the results of their search parameters.</p> 
    </section>

    <div class="line"></div>

    <section id="web-arch">
        <h3>Website Architecture</h3>
        <p> Figure 2. shows a summary of all the routes within the POPgen website and how they are linked. The Search page outputs the results page from user-input fields, such as the gene name, RS ID and Chromosome location. The results page then outputs the relevant data in tables and provides options to view the summary statistics within and between populations. Once the statistics options within the results page have been selected and searched, the statistics page will show the relevant statistics for the selected populations available within the database.</p>
        <img src="{{url_for('static', filename='help_files/web_arch.png')}}" align="middle" />
        <p style="font-weight: bold;"> Figure 2. POPgen web routes and links shown by solid black arrows. Web content shown by dotted lines. </p>
    </section>
    <div class="line"></div>

    <section id="run_popgen">
        <h3>Runnig POPgen</h3>
        
        <h4>How to run</h4>
        <p>POPgen has been deployed to the web successfully and the link can be found in the GitHub page. </p>
        <p>However, users can run POPgen locally. For POPgen to run, cloning the Web_Project repository from GitHub is needed.</p>
        <p>POPgen can run locally using:</p>
        <p>$ git clone https://github.com/Mani-varma1/Web_project</p>
        <p>$ cd Flask </p>
        <p>Virtual environment can be set up using: </p>
        <p>$ py -3 -m venv .venv</p>
        <p>$ .venv/Scripts/activate</p>
        <p>To install necessary packages:</p>
        <p>$ pip install -r packages.txt</p>
        <p>A script for populating the database has been provided for those that wish to set up a local database and wish to explore the dataset. The files required for generating the database have also been provided in the form of a zip file with an empty database. All that needs to be done is to run the database_populating.py script in the folder it is located in. This will extract the files to a new data folder and automatically populate the database. The script takes roughly 30 minutes to run. </p>
        <p>Finally, to run the application:</p>
        <p>$ python application.py</p>
        
        <h4>Walkthrough</h4>
        <p>The website lets you search by chromosome name and position (start-end), RS ID, gene name. It outputs SNPs information (chromosomal position, reference allele, alternate allele, frequencies) and allows you to select the desired statistics for calculation, based on the 5 populations provided. The populations can be selected as well. The statistics page contains the results of the calculated statistics (Table 1), as well as sliding windowed plots (Figure 3), of all statistics for better visualization and user experience.</p>
        <img src="{{url_for('static', filename='help_files/table_stats.png')}}" align="middle" />
        <p style="font-weight: bold;"> Table 1. Output of calculation of all statistics and all populations.</p>
        <img src="{{url_for('static', filename='help_files/plot_stats.png')}}" align="middle" />
        <p style="font-weight: bold;"> Figure 3. An example of a sliding plot for Tajimaâ€™s D using a search region of 18,000,000-19,000,000 for 5 populations, window size = 25,000 step size = 10,000.</p>
    </section>

    <div class="line"></div>

    <section id="technologies">        
        <h3>Technologies</h3>
        
        <h4>Flask</h4>
        <p>The structure of the website was defined using HTML tags/classes from the stylesheets of both Bootstrap and W3-CSS for each page within the website architecture. These HTML templates were coupled with CSS languages to produce a professional website design that provides a simple and user-friendly experience. The Bootstrap and W3-CSS tags create a framework for responsive elements to the website and help organize where each HTML component will appear on the website. For example, Bootstrap classes were used to center the team information on the about page. </p>
        
        <h4>HTML, CSS and JavaScript</h4>
        <p>HTML and CSS are the core elements of the frontend allowing for a well-organized and easy to understand user interface. HTML Document Object Models (DOM) were coupled with Bootstrap and W3-CSS libraries for creating a responsive and polished website. These libraries contain prebuilt custom classes for professional grade styling of our website. Although JavaScript is a core component of any modern website, we have used pre-existing libraries like JQuery to improve the interactivity of our website. </p>
        
        <h4>Forms and Validation</h4>
        <p>We often require input data types to be in specific format for the downstream analysis to work. User input can often be unpredictable and error prone. It is crucial to provide the user with appropriate feedback when mistakes are made. Although HTML has such attributes, they are limited in nature without extensive use of JavaScript. Extensions such as Flask WTforms, allow developers to easily create forms in an object-oriented way, which can be passed as a variable to the necessary templates. Setting up flask forms is extremely simple. The main advantages of using WTforms is their data validation, which allows the web server to check the input data and is within the constraints set by the application, and cross site request forgery form (CSRF) protection, which prevents malicious attacker to induce users to perform actions that they do not intend to perform. We can also easily create our own custom validation checks that are specific to our website forms without JavaScript. Having form validation can be coupled to CSS with Jinja2 to provide visual feedback to the users on their input errors, which can significantly improve the user experience and allows developers to focus more on the downstream analysis without formatting issues.</p>
        
        <h4>Sessions</h4>
        <p>Web apps often require data to be stored between each request as a user interacts with the app. However, HTTP is a stateless protocol and has no knowledge of other requests that were executed. This data could be stored in cookies which are stored on the user's computer. Due to security concerns and storage limit of 4KB, sessions are often used. Sessions are files that contain information on non-user specific requests they have executed. This allows us to access this data in multiple routes, without having to request user input for the same data on different pages. We use sessions to store information that was retrieved from queries which include SNP information and population specific frequencies. Prior to storage we must first serialize the data as session can only be stored in JSON format. Based on the user input parameters, the data size retrieved from the database could exceed 4KB. Therefore, we decided to store the sessions on the file systems, with no limits mentioned in the documentation. These sessions will automatically clear when either the browser is closed or replaced with data from a new query search if the user decides to change their search parameters. We can also have access to the session data within Jinja2 to create dynamic route links. Sessions can be used by JavaScript to perform various tasks in the front end without relying on the backend. Having sessions allows the user to quickly load the data if they close the tab by accident or navigate around the website, without having to search the database again.</p>
        
        <h4>Data Compression</h4>
        <p>Due to the amount of data within the VCF file it is not practical to store the raw genotype data for everyone as shown in VCF format. Doing so could significantly increase the storage requirements and deteriorate the querying speed. The repetitive nature of our data means there is lot of redundant data that violates the first normal form of databases, so we have decided to compress the data.  Although many compressions technologies exist, we took inspiration from Run length encoding (RLE) to create a custom compression method which retains the positions of the sample and minimises the data storage requirement. Allowing us to be more flexible in the future if we decide to include other statistical methods.</p>       
    </section>

    <div class="line"></div>  

    <section id="structure_design">
        <h3>Structure and Design</h3>
        <p>The structure of the website was defined using HTML tags/classes from the stylesheets of both Bootstrap and W3-CSS for each page within the website architecture. These HTML templates were coupled with CSS languages to produce a professional website design that provides a simple and user-friendly experience. The Bootstrap and W3-CSS tags create a framework for responsive elements to the website and help organize where each HTML component will appear on the website. For example, Bootstrap classes were used to center the team information on the about page. </p>
    </section>

    <div class="line"></div>  

    <section id="deployment">
        <h3>Deployment</h3>
        <h4><strong> Web Deployment </strong></h4>
        <p>Elastic Beanstalk was used from the Amazon Web Services (AWS).</p>
    </section>

    <div class="line"></div>    

    <h2>Data</h2>

    <div class="line"></div>
    
    <section id="collection">
        <h3>Data Collection</h3>
        <p>The data used to populate the database was primarily retrieved from the 1000 Genomes Project and the NCBI FTP server. The data used for this project was sourced from phase 3 data collection and this was done for two primary reasons:</p>
        <p>Phase three data contained information about both indels and multiallelic SNPs, which can be filtered to create a dataset only containing biallelic SNPs. Earlier phases of the project were not able to distinguish and identify multiallelic SNPs and would record them as separate entries within the VCF files (1KGP consortium, 2015).</p>
        <p>Phase three data used a larger sample size with more populations included. </p>
        <p>Phase three data used genomic positions based on the ensembl GrCh37.p19 reference build. This is important to note as the current RS ID's found on the dbSNP from ensembl have positions that correspond to both the GrCH37.p19 and the newer GrCH38.p19 reference build. The reasoning for choosing the GrCH37 build is explained in the Data Wrangling section.  </p>
        <p>Sample data was also required to allow sub-setting based on population, which was also obtained from 1KGP from the respective data collection page.</p>
        <p>Information regarding RS ID corresponding to position as well as gene consequence can be obtained from NCBI (Sherry et al, 2001). A VCF for the entire genome would need to be downloaded and then filtered. These files can be found here:</p>
        <p>https://ftp.ncbi.nih.gov/snp/latest_release/VCF/. </p>
        <p>Both GrCH37 (GCF_000001405.25) and GrCH38 (GCF_000001405.39) builds can be found with their corresponding tabix files. The chromosome fields are populated according RefSeq annotation. </p>
        <p>Data regarding gene names which corresponds to RS IDs can be obtained through several ways.  NCBI provides a command line tool which can retrieve information from dbSNP using their API (Kans, 2022). This is limited to 3 requests per second per IP address. The command used to retrieve gene names is shown below:</p>
        <p><code>for i in `cat x`; do esearch -db snp -query "${i} AND human [orgn]" | esummary | xtract -pattern DocumentSummary -element SNP_ID,NAME | uniq; done > output.txt</code></p>
    </section>  

    <div class="line"></div>

    <section id="pre-processing">
        <h3>Data Wrangling & Pre-processing</h3>
        <p>To keep storage requirements low, the database was constructed using VCF data from chromosome 22 for five populations: Mexican (MXL), British (GBR), Japanese (JPT), Punjabi (PJL), and Yoruba (YRI). Each of these populations belong to separate superpopulations, therefore information regarding superpopulations was left out as it would make the search options redundant for this build.</p>
        <p>Bcftools was used to subset only for biallelic SNPs and the 5 populations. The VCF files available from 1KGP contain empty fields for ID. To populate the database correctly, these would need to be filled. The following command from bcftools was used assuming we have a VCF file with annotated IDs and positions according to the same reference build:</p>
        <p><code>bcftools annotate -a 'path/to/dbsnp.vcf.gz' -c ID 'path/to/1KGP.vcf.gz' -o 'path/to/output.vcf.gz </code></p>
        <p>The reference VCF would need to be first filtered for only the chromosome of interest, which can also be done in bcftools. </p>
        <p>Some positions found on the VCF file which had previously been associated with a RS ID were no longer supported as of the current dbSNP build. As RS ID is required for the database, these SNPs were filtered from the final VCF file. This has led to gaps within the chromosome region, which may affect expected statistical values when searching these regions. These regions are: 20706257-20708536 & 25722272-25742171</p>
        <p>To get the data into a format that would make it easier to populate the database, the scikit-allel package in python was used (Miles & Harding, 2021). This has the advantage of using local storage array solutions to reduce load on memory usage. It also has many features to subset data and read it into custom array-like structures that work in a similar fashion to numpy arrays. The script used to wrangle this data can be found on the GitHub repository.</p>
        <p>This script will produce 6 tsv files; one query file and 5 population files. The population data contains genotype arrays and pre-calculated counts for genotypes and alleles per SNP.</p>
        <p>The query file requires further wrangling to get gene names inputted. A txt file containing gene names and RS IDs were then loaded into python as pandas dataframes along with the query file and merged on RS ID to produce a final query file. This file is provided along with population data, which can be used to create the database.</p>
    </section>  
    
    <div class="line"></div>

    <section id="storage">
        <h3>Data Storage</h3>
        <p>SQLite3 and SQLAlchemy were used to create and populate the database. Structures for the tables can be found in the models.py file. All search queries are initially made with the query table, which contains information about the chromosomal position, reference allele, alternate allele, and, where it is applicable, any genes that are implicated. Due to the unique property of RS IDs from the dbSNP, these were used as primary keys for which any searches would need to be matched to. The results page outputs the data stored within the query _search table (as seen in Figure 3). The other tables correspond to each population and contains information about the genotype frequency, allele frequency and phased genotype arrays for each SNP, with RS ID acting as a foreign key referencing the query_ search table. Both genotype frequency and allele frequency are stored in dictionary format within each of these fields.</p>
        <img src="{{url_for('static', filename='help_files/popgen_database_diagram.png')}}" align="middle" />
        <p style="font-weight: bold;"> Figure 4. POPgen database schematic: Shown is 6 tables and the links represented by solid lines </p>
    </section>  
    
    <div class="line"></div>   

    <section id="search">
        <h2>Search</h2>
        
        <p>Querying was done through SQLAlchemy. Searches can be made based on chromosomal positions, RS ID, and gene names (Using HGNC names). Individual SNPs can be queried by passing a single position in the start field or 1 RS ID. Multiple SNPs can be queried by providing a chromosomal region (start and stop positions) or a list of comma-separated RS IDs or gene names. Gene names have been limited to a maximum of five. </p>
        <p>Queries must be submitted in the correct format (especially for multiple SNP searches) and in the case of gene names, are case sensitive. </p>
        <p>NOTE: Large search queries can be made however this is dependent on hardware limitations. Significant slowdowns in page responsiveness can occur if 100,000s of SNPs are queried due to browser limitations. Keep this in mind when searching the database. On average, it takes ~1.5 minutes to run a search regardless of number of SNPs queried. This can take longer depending on how many SNPs need to be outputted to the page. It can take around 3 minutes to output a 500,000bp search.</p>
    </section>

    <div class="line"></div>

    <h2>Statistics</h2>

    <div class="line"></div> 

    <section id="sum-stats">
        <h3>Summary Statistics</h3>
        <p>Statistics analyses are performed to see whether there are significant differences between and within populations, using SNP information and genomic regions. All but one statistical values are calculated through Scikit-allel (Miles & Harding, 2021). </p>
        <p>Several summary statistics are available on POPgen. The user can select the summary statistics of interest, and values and plots will be outputted. The download button can be used to download summary statistics.</p>
        <p>Window size and step size are determined by the user. Step size is optional, and an empty field will result in window plots with no overlaps. Choosing an appropriate window size with respect to the size of the region searched is important to obtain meaningful analysis.</p>
        <p>Window sizes that are two small may result in NaN values due to insufficient number of SNPs in a window. The plots will still be produced, but with breaks where there are missing values. If this occurs, you may need to adjust the window and step size to reduce the occurrence of NaN values. Typically, a window size of 25,000 and step size of 10,000 is recommended for a search of a 100,000 bp region.</p>
        <p>As mentioned previously, there are two regions that may cause unexpected NaNs; 20706257-20708536 and 25722272 â€“ 25742171. These regions contain missing SNPs found in the original VCF due to unsupported RS ID. Care needs to be taken when searching these regions, especially the latter. </p>
        
        <h4>Observed Homozygosity</h4>
        <p>This is a measure of the total sum of homozygous reference and homozygous alternative for each SNP. The value outputted indicates the proportion in which the SNP is homozygous, and a high value indicates the likelihood of the SNP being homozygous. This is based on a direct count method highlighted in Sabatti & Risch, 2002.</p>
        <img src="{{url_for('static', filename='help_files/Obs_Homo.png')}}" align="middle" />
        
        <h4>Genetic Diversity</h4>
        <p>Genetic Diversity can be used to see whether there is variation between selected populations. </p>
        <p>Nucleotide Diversity</p>
        <p>Nucleotide Diversity ( Ï€ ) is used as a measure of genetic diversity on POPgen. This is defined as the average number of nucleotide differences per site when two sequences are chosen randomly from the sample population. The formula outputs Ï€ and a high Ï€ indicates high genetic variation.</p>
        <img src="{{url_for('static', filename='help_files/Nucl_Div.png')}}" align="middle" />
        <p>Here, xi and xj refer to the population frequency of the ith and jth nucleomorph, and n refer to sample size. Ï€ ij refers to the differences in nucleotides between the ith and jth sequences (Nei & Li, 1979). </p>
        
        <h4>Haplotype Diversity</h4>
        <p>Haplotype Diversity, also known as gene diversity, can be defined as the probability of two alleles randomly being different when sampled (de Jong et al., 2011). </p>
        <img src="{{url_for('static', filename='help_files/Haplo_Div.png')}}" align="middle" />
        <p>The function estimates haplotype diversity (H), where xi refers to the relative haplotype diversity in the sample and N refers to the sample size.  A value close to 1 indicates that the probability of picking two haplotypes which are the same are extremely high. A value close to 0 indicates that there are lots of haplotypes indicating more variation. </p>
        
        <h4>Neutrality Test</h4>
        <p>Neutrality tests are used in population genetics to assess the goodness-of-fit of standard neutral model (Achaz, 2009) based on the frequency spectrum such as Tajimaâ€™s or Fu and Liâ€™s F.</p>
        <p>Tajimasâ€™ D</p>
        <p>Tajimaâ€™s D (D) is used as a test against neutrality, which is calculated through genotype arrays and counting the alleles. The formula compares the difference in values of Watterson theta (ratio of number of SNPs observed to expected number of SNPs based on neutral model of evolution) and expected number of segregating sites (according to a neutral model of evaluation). </p>
        <img src="{{url_for('static', filename='help_files/TajimasD.png')}}" align="middle" />
        <p>Interpreting D</p>
        <p>D = 0: The observed and expected variation are similar. One conclusion is there is no evidence of selection for the observed sequence.</p>
        <p>D > 0: Observed variation is greater than expected. This may indicate that the alternate alleles are scarce and that there may be a balancing selection. </p>
        <p>D < 0: Observed variation is less than expected. This may indicate that the alternate alleles are abundant, indicating possibility of a selection pressure/selective sweep.</p>
        <p>With single values, it may be hard to ascertain the meaning of these values, as these interpretations are only relevant if the value for D is statistically significant. The sliding window feature can obtain a range of values along an overall searched region, which can then be used to identify regions that may be significantly different from other observed values. A table of confidence limits for D can be found in the original published paper, which can be used to determine significance of individual observed values. (Tajima, 1989).</p>
        
        <h4>Population Genetic Variation</h4>
        <p>Fst</p>
        <p>Fst is a measure of genetic variance which is explained by population structure. POPgen calculates the Fst values for each pair of populations the user has selected. The function is calculated through the method of Hudson (1992). Values range from 0 (low degree of differentiation) to 1 (High degree of differentiation).</p>
        <img src="{{url_for('static', filename='help_files/Fst.png')}}" align="middle" />
        <p>Here, P1 refers to sample allele frequency in the first population and P2 refers to sample allele frequency in the second population. The function used by POPgen produces the numerator (which defines the mean number of differences within a population) and the denominator (defines the mean number of differences between populations). The sum of the numerators is divided by the sum of the denominators to obtain an average value for Fst over all variants searched. This definition of Fst was chosen as it is independent of sample size and is recommended when doing pair-wise comparisons (Bhatia et al, 2013). </p>
    </section>  
    
    <div class="line"></div> 

    <section id="plots">
        <h3>Plots</h3>
        <h4>Sliding Window Plots</h4>
        <img src="{{url_for('static', filename='help_files/sliding_window_plots.png')}}" align="middle" />
        <p style="font-weight: bold;">Figure 5. An example of a sliding plot for Fst using a search region of 18,000,000-19,000,000 for 5 populations, window size = 25,000 step size = 10,000. </p>
        <p>A sliding window plot is created using a window of a fixed size which is sliding along the genomic region of interest. The sliding window will move in steps of a specified size from the beginning to end position. At each window, the statistical values are calculated and plotted against the average genomic position.</p>
        <p>A sliding window plot is calculated and plotted using positions, population data, window sizes and step sizes. Window and step sizes are given in base pairs and can be specified by the user.</p>
        <p>Plots are created using Plotly package (Nucleotide Diversity, Haplotype Diversity, Tajimaâ€™s D and Fst). The different populations can be selected either using the lines to the side of the plot.</p>
    </section>  
    

    <div class="line"></div> 


    <h2>Limitations</h2>


    <div class="line"></div> 

    <section id="dataset_limitations">
        <h3>Dataset Limitations</h3>
        <p>The dataset itself has issues, and no single dataset from the 1KGP was perfect. As the nature of this project involves using datasets that are constantly being updated themselves, there are some discrepancies with the final form of the dataset used. Recent dbSNP builds no longer supports certain SNPs that are found in the dataset used. To satisfy the requirements for the database, these SNPs would need to be removed during the data wrangling process. This will inevitably lead to regions in the VCF files that hold no SNPs. One potential workaround would be to include the non-supported RS IDs as they are still searchable on the dbSNP (the individual pages simply mention that its unsupported), however this would require mixing different builds for dbSNP to populate the missing field and using several reference files, to ensure that no positions were left unannotated. As bcftools was used to annotate the RS IDs, this can become complicated as it tends to override previous annotations. </p>
    </section>
  
    <div class="line"></div> 

    <section id="database_limitations">
        <h3>Database Limitations</h3>
        <p>The format of the tables, regarding populations, is unnormalised. It would be best practice, especially with larger datasets spanning multiple chromosomes and more populations, to split the genotype and allele frequency into four or five separate columns and storing only 1 value in each field. This would allow the database to satisfy the first normal form (1NF). Another issue is regarding the formatting of phased genotype arrays. Initially, two additional tables were trialed, one table containing sample IDs and information regarding each sample, and another table containing genotype array data, with a column containing RS ID and sample ID. The issue with this approach was the number of rows that would need to be populated for a single chromosome. For chromosome 22 and 500 samples, this would equate to ~500,000,000 rows. For the purposes of this version of the application, keeping array data in the form that we used can simplify the populating of the database, but changes to individual genotypes would prove to be difficult.</p>
    </section>
            
    <div class="line"></div> 

    <section id="tech_limitations">
        <h3>Technologies Limitations</h3>
        <p>Flask is only viable for a small-scale application. More complex applications need support for more built-in packages and libraries. Consequently, there could be an issue when scaling the application. Django would be a good alternative option, as it offers everything Flask has and more, but with a steeper learning curve.</p>
        <p>Most web apps use HTML, CSS and JS. Although we used some built in JS libraries, the use of these was very limited. There is a need for more custom JS. For a full control of our website, use JS with libraries such as react, which also allows us to build our own API and make the website more robust.</p>
        <p>Storage requirements could be an issue if multiple users access the database and search for large regions at the same time.</p>
    </section>

    <div class="line"></div> 

    <section id="future_developments">
        <h2>Future Developments</h2>
        <p>1. Expanding dataset to include entire human genome â€“ significant storage requirements and data wrangling required. It took around 1 week to get the gene names for the smallest autosome. The entire dbSNP database would need to be accessed to generate gene names for each RS ID. Also, a more efficient method for wrangling the data would need to be employed as the current scripts can take up to 1 hour to go from VCF files to a populated database.  </p>
        <p>2. Expand to all populations from IGSR and allow for filtering by superpopulations. Would need to create additional tables for each population and use dictionaries to categorise populations by superpopulations.  </p>
        <p>3. More meaningful stats options (e.g. PCA) using genotype array data â€“ would require restructuring of database to separate 1 genotype per field. Also, would require sample information table for information about sample ID, population and superpopulation.  </p>
        <p>4. Exploring options of non-relational databases that may be better suited for the type of data we are using. For example, multivalue databases or object databases could be used for data that was stored as lists or dictionaries. </p>
        <p>5. Improvements to algorithms for better efficiency â€“ Code relies on some nested for loops that can significantly increase memory usage and time complexity. </p>
        <p>6. Cleaner code â€“ streamline repetitive elements by using more classes and functions.</p>
          
    </section>
            
 {% endblock content %}
